src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitSuperClassTest.java

```diff
@@ -0,0 +1,165 @@
+        "com.puppycrawl.tools.checkstyle.checks.metrics.ClassFanOutComplexityCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.AbstractAccessControlNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.CatchParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.ClassTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.ConstantNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.IllegalIdentifierNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.InterfaceTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LambdaParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LocalFinalVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.LocalVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.MemberNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.MethodNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.MethodTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.ParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.PatternVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordComponentNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.StaticVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.TypeNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.TypecastParenPadCheck"
+    );
+
+    /**
+     * ArchCondition checking that a class has a particular super class.
+     *
+     * @param superclass the superclass
+     * @return ArchCondition checking that a class has a particular super class
+     */
+    private static ArchCondition<JavaClass> haveSuperclass(Class<?> superclass) {
+        return new SuperclassArchCondition(superclass);
+    }
+
+    /**
+     * Tests that all checks either extend {@link AbstractCheck} or {@link AbstractFileSetCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testChecksShouldExtendAbstractCheckOrAbstractFileSetCheck() {
+        final JavaClasses checksPackage = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.checks");
+
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition<>(
+            SUPPRESSED_CLASSES, "be suppressed");
+
+        final ArchRule checksShouldHaveCorrectSuperclass = classes()
+            .that()
+            .haveSimpleNameEndingWith("Check")
```

Comment:
Again I think `ModuleReflectionUtil` should be a consideration here.

Reply:
Is it possible to reuse
https://github.com/checkstyle/checkstyle/blob/78584f79f43a3b788a7d7d23203463b506f5ef5a/src/main/java/com/puppycrawl/tools/checkstyle/utils/ModuleReflectionUtil.java#L57

special class provider.

Reply:
Or our specific preticate
https://github.com/checkstyle/checkstyle/blob/78584f79f43a3b788a7d7d23203463b506f5ef5a/src/main/java/com/puppycrawl/tools/checkstyle/utils/ModuleReflectionUtil.java#L75

Reply:
Done.

Reply:
Partially done, but opened up a new comment below.